internal/network/wrapper.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: Copyright (c) 2023, CIQ, Inc. All rights reserved
+// SPDX-License-Identifier: Apache-2.0
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/jasonyangshadow/apptheus/internal/cgroup/parser"
+	"github.com/jasonyangshadow/apptheus/internal/monitor"
+	"github.com/jasonyangshadow/apptheus/storage"
+	"github.com/prometheus/exporter-toolkit/web"
+	"toolman.org/net/peercred"
+)
+
+type ServerOption struct {
+	Server      *http.Server
+	WebConfig   *web.FlagConfig
+	MetricStore storage.MetricStore
+	Logger      log.Logger
+	SocketPath  string
+	TrustedPath string
+	Interval    *time.Ticker
+	ErrCh       chan error
+}
+
+type WrappedInstance struct {
+	*parser.ContainerInfo
+	*monitor.Instance
+	net.Conn
+}
+
+type WrappedListener struct {
+	*peercred.Listener
+	TrustedPath  string
+	Option       *ServerOption
+	ContainerMap map[string]*WrappedInstance
+}
+
+func (l *WrappedListener) Accept() (net.Conn, error) {
+	conn, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	pid := conn.(*peercred.Conn).Ucred.Pid
+
+	// verification by pid
+	link, err := os.Readlink(fmt.Sprintf("/proc/%d/exe", pid))
```

Comment:
At some point we might consider to use https://man7.org/linux/man-pages/man2/pidfd_open.2.html for polling

Reply:
hmm, can you explain more on this comment?
I checked the doc, it looks like we can use this pidfd_open to monitor whether the apptainer instance exits, rather than checking whether there are processes inside cgroup. 
Regarding the verification of trust path, I can only get the value of `anon_inode:[pidfd]` by reading the link.

Reply:
To avoid race or pid recycle issue, it would look like this
```
package main

import(
    "fmt"
    "golang.org/x/sys/unix"
)

func main() {
    pid := unix.Getppid()

    pidfd, err := unix.PidfdOpen(pid, 0)
    if err != nil {
        panic(err)
    }

    dirfd, err := unix.Open(fmt.Sprintf("/proc/%d", pid), unix.O_DIRECTORY, 0)
    if err != nil {
        panic(err)
    }

    if err := unix.PidfdSendSignal(pidfd, 0, nil, 0); err != nil {
        // process died or pid has been recycled for another process
        panic(err)
    }

    buf := make([]byte, 4096)
    n, err := unix.Readlinkat(dirfd, "exe", buf)
    if err != nil {
        panic(err)
    }

    fmt.Println(string(buf[:n]))
}
```